cirq-core/cirq/experiments/parallel_readout_calibration.py

```diff
@@ -0,0 +1,136 @@
+        data was finished being collected from the sampler.  Note that,
+        if there did not exist a trial where a given qubit was set to |0〉,
+        the zero-state error will be set to `nan` (not a number).  Likewise
+        for qubits with no |1〉trial and one-state error.
+    """
+    qubits = list(qubits)
+
+    if trials <= 0:
+        raise ValueError("Must provide non-zero trials for readout calibration.")
+    if repetitions <= 0:
+        raise ValueError("Must provide non-zero repetition for readout calibration.")
+    if bit_strings is None:
+        bit_strings = [random.getrandbits(trials) for _ in qubits]
+    if len(bit_strings) != len(qubits):
+        raise ValueError(
+            f'If providing bit_strings, # of bit strings ({len(bit_strings)}) '
+            f'must equal # of qubits ({len(qubits)})'
+        )
+
+    all_circuits = []
+    all_sweeps: List[study.Sweepable] = []
+    if trials_per_batch is not None:
+        if trials_per_batch <= 0:
+            raise ValueError("Must provide non-zero trials_per_batch for readout calibration.")
+        num_batchs = trials // trials_per_batch
+        if trials % trials_per_batch > 0:
+            num_batchs += 1
+    else:
+        num_batchs = 1
+        trials_per_batch = trials
+    for batch in range(num_batchs):
+        circuit = circuits.Circuit()
+        single_sweeps = []
+        for idx, q in enumerate(qubits):
+            sym_val = f'bit_{idx}'
+            current_trial = bit_strings[idx]
+            trial_range = range(batch * trials_per_batch, (batch + 1) * trials_per_batch)
+            circuit.append(ops.X(q) ** sympy.Symbol(sym_val))
+            single_sweeps.append(
+                study.Points(
+                    key=sym_val, points=[(current_trial >> bit) & 1 for bit in trial_range]
+                )
+            )
+
+        circuit.append(ops.measure_each(*qubits, key_func=repr))
+        total_sweeps = study.Zip(*single_sweeps)
+        all_circuits.append(circuit)
+        all_sweeps.append(total_sweeps)
+
+    results = sampler.run_batch(all_circuits, all_sweeps, repetitions=repetitions)
+    timestamp = time.time()
+
+    zero_state_trials: Dict[cirq.Qid, List[float]] = {q: [] for q in qubits}
+    one_state_trials: Dict[cirq.Qid, List[float]] = {q: [] for q in qubits}
+    for batch_result in results:
+        for trial_idx, trial_result in enumerate(batch_result):
+            for idx, q in enumerate(qubits):
+                had_x_gate = (bit_strings[idx] >> trial_idx) & 1
+                if had_x_gate:
+                    one_state_trials[q].append(1 - np.mean(trial_result.measurements[repr(q)]))
+                else:
+                    zero_state_trials[q].append(np.mean(trial_result.measurements[repr(q)]))
+
+    zero_state_errors = {q: np.mean(zero_state_trials[q]) for q in qubits}
+    one_state_errors = {q: np.mean(one_state_trials[q]) for q in qubits}
```

Comment:
I'm worried this math might have some wiggle room for errors. If I break up some batches into the following sizes:
`num_batches = 3`
and batch_sizes of `[10, 10, 2]`
Then this code will weight each of the batch sizes equally and increase the weight of the final 2 samples disproportionately to the original two 10 entries. Am I missing something here ?

Reply:
This is a good catch!  I _think_ it may have been okay since the repetitions are the same for each trial, but I have fixed this to count up the trials anyway, since I think the previous code was error prone if not outright broken.